packages/toolpad-app/src/server/auth.ts

```diff
@@ -44,12 +84,31 @@ export function createAuthHandler(base: string): Router {
         trustHost: true,
         callbacks: {
           async signIn({ account, profile }) {
+            const dom = await project.loadDom();
+
+            const app = appDom.getApp(dom);
+            const requiredEmails = app.attributes.authentication?.requiredEmail ?? [];
+
+            if (account?.provider === 'github') {
+              return Boolean(
+                profile?.verifiedEmails &&
+                  profile.verifiedEmails.length > 0 &&
+                  (requiredEmails.length === 0 ||
+                    requiredEmails.some((requiredEmail) =>
+                      profile.verifiedEmails!.some((verifiedEmail) =>
+                        new RegExp(requiredEmail).test(verifiedEmail),
+                      ),
+                    )),
+              );
+            }
             if (account?.provider === 'google') {
               return Boolean(
                 profile?.email_verified &&
                   profile?.email &&
-                  (!process.env.TOOLPAD_GOOGLE_AUTH_DOMAIN ||
-                    profile.email.endsWith(`@${process.env.TOOLPAD_GOOGLE_AUTH_DOMAIN}`)),
+                  (requiredEmails.length === 0 ||
+                    requiredEmails.some(
+                      (requiredEmail) => new RegExp(requiredEmail).test(profile.email!) ?? false,
```

Comment:
I'm not a huge fan of using regexes for this. Seems a bit like a footgun to me. e.g. users forgetting to add `$` on the end, or forgetting to escape `.`. Wouldn't >95% of the usecases be covered by just restricting the email domain? Can you benchmark other products on how configurable this is?

Reply:
Ok I'll give a look into how other tools do it, but last case scenario we can just check if the user email ends with the specified domain preceded by @, instead of regex.